### ISSUE TITLE
fit_generator - how accuracy is reported

### ISSUE THREAD

**Issue by mjs-wpi** [fchollet/keras#5432]
I'm training a 5-class classification model that uses a generator to balance batches during training; every batch contains the same number of examples from each class. 

My Accuracy on the Training Set is 80% and my accuracy on the Validation Set is 35%. Here is the problem: when I construct a confusions matrix on the Training Set, accuracy appears to be way less than what the verbose=1 flag is reporting as my accuracy (confusion matrix = 25%).

My question is what could be going wrong? Why is the accuracy on the training set _during_ training different from accuracy reported on the training set _in the final confusion matrix_? Is this a problem with my generator?

```python
def training_gen(batchsize = 100):

    lengths = {}
    order = {}
    index = {}
    for i in ['S1_64.npz', 'S2_64.npz', 'S3_64.npz', 'R_64.npz', 'W_64.npz']:
        with np.load('./' + i) as data:
            lengths[i] = data['Y'].shape[0]
            order[i] = np.random.permutation(lengths[i]) - 1
            index[i] = 0

    xbatch_dict = {}
    ybatch_dict = {}

    while 1:
        for i in ['S1_64.npz', 'S2_64.npz', 'S3_64.npz', 'R_64.npz', 'W_64.npz']:
            with np.load('./' + i) as data:
                xbatch_dict[i] = data['X'][order[i][index[i]:index[i] + int(batchsize/5)],:,:]
                ybatch_dict[i] = data['Y'][order[i][index[i]:index[i] + int(batchsize/5)],:]
                index[i] += int(batchsize/5)

        xbatch = np.vstack(list(xbatch_dict.values()))
        ybatch = np.vstack(list(ybatch_dict.values()))
        yield xbatch, ybatch                                                                                                                                                    

        del (xbatch, ybatch)

        flag = 0
        for i in ['S1_64.npz', 'S2_64.npz', 'S3_64.npz', 'R_64.npz', 'W_64.npz']:
            if index[i] >= max(lengths.values()):
                flag = 1
                break
            if index[i] + int(batchsize/5) - 1 > lengths[i]:
                index[i] = 0

        if flag == 1:
            for i in ['S1_64.npz', 'S2_64.npz', 'S3_64.npz', 'R_64.npz', 'W_64.npz']:
                index[i] = 0
                order[i] = np.random.shuffle(order[i])



```

**Comment by patyork** [comment]
Mostly because the model is being trained; the model at "Batch 1/100" will more than likely be markedly different than the model at "Batch 100/100".

During training, the accuracy is being averaged as each batch is seen, evaluated, and updating the model.

**Comment by mjs-wpi** [comment]
@patyork, thanks for your answer, but I'm still confused. Even at "Batch 100/100" The accuracy seems very high:

> Epoch 1/5
3000/3000 [==============================] - 3876s - loss: 2.9595 - acc: 0.6273 - val_loss: 1.8416 - val_acc: 0.3393
Epoch 2/5
3000/3000 [==============================] - 3800s - loss: 2.7473 - acc: 0.7207 - val_loss: 1.8439 - val_acc: 0.3323
Epoch 3/5
3000/3000 [==============================] - 3842s - loss: 2.6457 - acc: 0.7573 - val_loss: 1.8546 - val_acc: 0.3431
Epoch 4/5
3000/3000 [==============================] - 3819s - loss: 2.5409 - acc: 0.7883 - val_loss: 1.8698 - val_acc: 0.3481
Epoch 5/5
3000/3000 [==============================] - 3791s - loss: 2.4950 - acc: 0.8020 - val_loss: 1.8604 - val_acc: 0.3546

**Comment by patyork** [comment]
```
Epoch 1/5
3000/3000 [==============================] - 3876s - loss: 2.9595 - acc: 0.6273 - val_loss: 1.8416 - val_acc: 0.3393
```
This is saying that, over the epoch, the average accuracy was 62.7% - some batches were higher, some were lower, but on average the model correctly predicted 62.7% of the samples. After the epoch (3000 examples), the validation set is being ran, and its showing an average accuracy of 33.9%.

**Comment by patyork** [comment]
Based on your file naming, im going to assume you've got 3 classes; same number of samples for each. In your generator your taking batch/5 samples from each file. You've got 3 files labeled S, meaning a batch has 3/5 S labels. Your model is therefore learning to always output S, and is getting about 3/5 or 60% correct during training. However, during validation, S only makes up 33% or so of your data. This explaining your confusion.

Fix your generator.

-----Original Message-----

**Comment by mjs-wpi** [comment]
Hi @patyork, thank you again for looking at the code. I actually have 5 classes; S1, S2, and S3 are different labels, "sleep stage 1," "steep stage2," and "sleep stage 3 (deep sleep)." So this is still not explaining why I'm seeing 60% average accuracy over an epoch but not in the confusion matrix.
